src/main/java/com/github/msemys/esjc/tcp/handler/IdentificationHandler.java

```diff
@@ -40,7 +40,16 @@ public IdentificationHandler(String connectionName, Duration timeout) {
 
     @Override
     protected void channelRead0(ChannelHandlerContext ctx, TcpPackage msg) throws Exception {
-        if (timeoutTask != null && !timeoutTask.isDone() && msg.correlationId.equals(correlationId)) {
+
```

Comment:
maybe its better to lock entire method (better code readability):
```java
protected void channelRead0(...) {
    synchronized (timeoutTaskLock) {
        if (timeoutTask != null && !timeoutTask.isDone() && msg.correlationId.equals(correlationId)) {
            ...
        } else { 
            ... 
        }
    }
}
```

Reply:
Wouldn't executing `complete` or `cancelTimeoutTask` under `timeoutTaskLock` be a bit dangerous in regards to locking itself?

Reply:
its reentrant lock - a thread can acquire a lock that it already owns

Reply:
I will add a new commit.

Reply:
@msemys I just looked at the `complete` method again, and please correct me if I am wrong, but it seems that  there is no guarantee that completion handler will not invoke something that is going to manipulate channel read. Same is true for the `ctx.fireChannelRead(msg)`. if any of these trigger some non-Future-proofed read from another thread, it might lock.

On top of that, the scope of the lock will become *much* wider, and this is exactly what I was trying to avoid.

I have no deep knowledge of your codebase, so please forgive me if I am missing something very obvious. I was (and still am) looking for correctness of the code, but I agree - code readability is important too. Do you still want me to extend the locking scope?

Reply:
your code looks safer :+1: 

about readability... maybe we could do it in this way?
```java
protected void channelRead0(...) {
        final boolean isClientIdentificationResponse;

        synchronized (timeoutTaskLock) {
            isClientIdentificationResponse = (timeoutTask != null) && !timeoutTask.isDone() && msg.correlationId.equals(correlationId);
        }

        if (isClientIdentificationResponse) {
            switch (msg.command) {
                ...
                ...
                ...
```

Reply:
This is cleaner and it should work actually, since we are not using any of these variables after that `if` statement anyway.  I will change that commit and rebase on current master.

FWIW, the submitted patch is in the production for 4 month already, and we do not get any "funny" stack traces anymore (fingers crossed).

Reply:
please take a look.

Reply:
:+1: